# asap2
Amplicon sequencing of marker genes such as 16S rDNA, 18S rDNA, ITS and others has been widely used to survey and characterize microbial communities in countless ecological and environmental studies. However, the complex data analyses have required many interfering manual steps often leading to inconsistencies in results. Here, we have developed a pipeline, amplicon sequence analysis pipeline 2 (ASAP 2), to automate and glide through the processes without the usual manual inspections and userâ€™s interference, for instance, in the detection of barcode orientation, selection of high-quality region of reads, and determination of resampling depth and many more. The pipeline integrates all the analytical processes such as importing data, demultiplexing, summarizing read profiles, trimming quality, denoising, removing chimeric sequences, and making the feature (ASV) table among others, using QIIME 2, Vegan and other tools. The pipeline accepts multiple file formats as input including multiplexed or demultiplexed, paired-end or single-end, barcode inside or outside and raw or intermediate data (e.g. feature table). The outputs include taxonomic classification, alpha/beta diversity, community composition, ordination analysis and statistical tests (variable selection, CCA, RDA, etc.). ASAP 2 supports merging multiple sequencing runs which helps integrate and compare data from different sources (public databases and collaborators). The pipeline minimizes hands-on interference and runs amplicon sequencing analysis automatically and consistently.
